packages/frontend/workspace/src/providers/sync/peer.ts

```diff
@@ -0,0 +1,386 @@
+import { DebugLogger } from '@affine/debug';
+import { Slot } from '@blocksuite/global/utils';
+import type { Doc } from 'yjs';
+import { applyUpdate, encodeStateAsUpdate, encodeStateVector } from 'yjs';
+
+import type { Storage } from '../storage';
+import { AsyncQueue } from '../utils/async-queue';
+import { throwIfAborted } from '../utils/throw-if-aborted';
+
+/**
+ * # SyncPeer
+ * A SyncPeer is responsible for syncing one Storage with one Y.Doc and its subdocs.
+ *
+ *                    ┌─────┐
+ *                    │Start│
+ *                    └──┬──┘
+ *                       │
+ *    ┌──────┐     ┌─────▼──────┐        ┌────┐
+ *    │listen◄─────┤pull rootdoc│        │peer│
+ *    └──┬───┘     └─────┬──────┘        └──┬─┘
+ *       │               │ onLoad()         │
+ *    ┌──▼───┐     ┌─────▼──────┐      ┌────▼────┐
+ *    │listen◄─────┤pull subdocs│      │subscribe│
+ *    └──┬───┘     └─────┬──────┘      └────┬────┘
+ *       │               │ onReady()        │
+ *    ┌──▼──┐      ┌─────▼───────┐       ┌──▼──┐
+ *    │queue├──────►apply updates◄───────┤queue│
+ *    └─────┘      └─────────────┘       └─────┘
+ *
+ * listen: listen for updates from ydoc, typically from user modifications.
+ * subscribe: listen for updates from storage, typically from other users.
+ *
+ */
+export enum SyncPeerStatus {
+  Stopped = 0,
```

Comment:
Is it possible that we make status more fine-grained? For example, give more observability on how many docs are being syncing (to calculate the syncing progress etc)

Reply:
year, sure. we gonna need a status map. I have the consideration in design

Reply:
This is not difficult, but requires some UI design to display this information, which will be implemented in the future